src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/FeatureSwitchManager.cs

```diff
@@ -0,0 +1,837 @@
+                        break;
+                }
+            }
+
+            // Now sweep unreachable basic blocks by replacing them with nops
+            bool hasUnmarkedIntructions = false;
+            foreach (var flag in flags)
+            {
+                if ((flag & OpcodeFlags.InstructionStart) != 0 &&
+                    (flag & OpcodeFlags.Mark) == 0)
+                {
+                    hasUnmarkedIntructions = true;
+                }
+            }
+
+            if (!hasUnmarkedIntructions)
+                return method;
+
+            byte[] newBody = (byte[])methodBytes.Clone();
+            int position = 0;
+            int endOfUsefulCode = 0;
+            while (position < newBody.Length)
+            {
+                Debug.Assert((flags[position] & OpcodeFlags.InstructionStart) != 0);
+
+                bool erase = (flags[position] & OpcodeFlags.Mark) == 0;
+
+                // If we need to erase this instruction and this instruction starts a visible basic
+                // block, we need to neutralize it by rewriting it into an infinite loop ("br $-2").
+                if (erase &&
+                    (flags[position] & OpcodeFlags.VisibleBasicBlockStart) != 0)
+                {
+                    // Make sure there's enough space (2 bytes) to neutralize the basic block.
+                    if (position + 1 < newBody.Length
+                        && (flags[position + 1] == 0 ||
+                        (((flags[position + 1] & OpcodeFlags.Mark) == 0
+                        && (flags[position + 1] & OpcodeFlags.VisibleBasicBlockStart) == 0))))
+                    {
+                        newBody[position] = (byte)ILOpCode.Br_s;
+                        newBody[position + 1] = unchecked((byte)-2);
+                        position += 2;
+                        endOfUsefulCode = position;
+
+                        // If we reached the end of the instruction or stream, loop over.
+                        // Otherwise continue down to nop out the remainder of this instruction.
+                        if (position == newBody.Length
+                            || flags[position] != 0)
+                            continue;
+                    }
+                    else
+                    {
+                        // We cannot neutralize the basic block, so better leave the method alone.
+                        return method;
+                    }
+                }
+
+                // Make sure to nop out the entire instruction
+                do
+                {
+                    if (erase)
+                        newBody[position] = (byte)ILOpCode.Nop;
+                    position++;
+                    if (!erase)
+                        endOfUsefulCode = position;
+                }
+                while (position < newBody.Length && flags[position] == 0);
+            }
+
+            // RyuJIT doesn't like when there's unreachable garbage at the end of the method.
```

Comment:
Where is the code that chokes on this? I would expect that RyuJIT should handle fine.

Reply:
IIRC it was this one:

https://github.com/dotnet/runtime/blob/f513349a68a2f0e05189ca2fe2972b1da306995e/src/coreclr/jit/flowgraph.cpp#L6021-L6024

RyuJIT seems to generate basic blocks also for junk bytes not referenced from any live code and this code chokes on it. It was a block of nops following a `br $-2`.

Reply:
I'll double check and capture an ngendump

Reply:
Yeah, it's that one:

```
****** START compiling System.SR:Format(System.String,System.Object):System.String (MethodHash=b9e74bbf)
Generating code for Windows x64
OPTIONS: compCodeOpt = BLENDED_CODE
OPTIONS: compDbgCode = true
OPTIONS: compDbgInfo = true
OPTIONS: compDbgEnC  = false
OPTIONS: compProcedureSplitting   = false
OPTIONS: compProcedureSplittingEH = false
OPTIONS: Jit invoked for ngen
IL to import:
IL_0000  28 01 01 00 06    call         0x6000101
IL_0005  2c 19             brfalse.s    25 (IL_0020)
IL_0007  72 ad 01 00 70    ldstr        0x700001AD
IL_000c  18                ldc.i4.2
IL_000d  8d 06 00 00 01    newarr       0x1000006
IL_0012  25                dup
IL_0013  16                ldc.i4.0
IL_0014  02                ldarg.0
IL_0015  a2                stelem.ref
IL_0016  25                dup
IL_0017  17                ldc.i4.1
IL_0018  03                ldarg.1
IL_0019  a2                stelem.ref
IL_001a  28 6d 00 00 0a    call         0xA00006D
IL_001f  2a                ret
IL_0020  2b fe             br.s         -2 (IL_0020)
IL_0022  00                nop
IL_0023  00                nop
IL_0024  00                nop
IL_0025  00                nop
IL_0026  00                nop
IL_0027  00                nop

lvaSetClass: setting class for V00 to (4000000000420430) System.String
Arg #0    passed in register(s) rcx

lvaSetClass: setting class for V01 to (4000000000420260) System.Object
Arg #1    passed in register(s) rdx

lvaGrabTemp returning 2 (V02 tmp0) (a long lifetime temp) called for OutgoingArgSpace.
; Initial local variable assignments
;
;  V00 arg0              ref  class-hnd
;  V01 arg1              ref  class-hnd
;  V02 OutArgs        lclBlk <na>  "OutgoingArgSpace"
*************** In compInitDebuggingInfo() for System.SR:Format(System.String,System.Object):System.String
getVars() returned cVars = 0, extendOthers = true
info.compVarScopesCount = 2
        VarNum  LVNum         Name      Beg     End
 0:     00h     00h       V00 arg0      000h    028h
 1:     01h     01h       V01 arg1      000h    028h
New Basic Block BB01 [0000] created.
New scratch BB01
Debuggable code - Add new BB01 [0000] to perform initialization of variables
info.compStmtOffsetsCount    = 0
info.compStmtOffsetsImplicit = 0007h ( STACK_EMPTY NOP CALL_SITE )
*************** In fgFindBasicBlocks() for System.SR:Format(System.String,System.Object):System.String
weight= 79 : state  40 [ call ]
weight= 27 : state  44 [ brfalse.s ]
weight= 66 : state 102 [ ldstr ]
weight= 34 : state  25 [ ldc.i4.2 ]
weight=152 : state 118 [ newarr ]
weight= 11 : state  38 [ dup ]
weight= 15 : state  23 [ ldc.i4.0 ]
weight= 10 : state   3 [ ldarg.0 ]
weight= 94 : state 139 [ stelem.ref ]
weight= 11 : state  38 [ dup ]
weight= 28 : state  24 [ ldc.i4.1 ]
weight= 16 : state   4 [ ldarg.1 ]
weight= 94 : state 139 [ stelem.ref ]
weight= 79 : state  40 [ call ]
weight= 19 : state  42 [ ret ]
weight= 44 : state  43 [ br.s ]
weight= 65 : state   2 [ noshow ]
weight= 65 : state   2 [ noshow ]
weight= 65 : state   2 [ noshow ]
weight= 65 : state   2 [ noshow ]
weight= 65 : state   2 [ noshow ]
weight= 65 : state   2 [ noshow ]
Jump targets:
  IL_0020
New Basic Block BB02 [0001] created.
BB02 [000..007)
New Basic Block BB03 [0002] created.
BB03 [007..020)
New Basic Block BB04 [0003] created.
BB04 [020..022)
```